pkg/jobstore/boltdb/store.go

```diff
@@ -0,0 +1,1013 @@
+			ids, err := b.tagsIndex.List(tx, tagLabel)
+			if err != nil {
+				return nil, err
+			}
+
+			for _, k := range ids {
+				jobSet[string(k)] = struct{}{}
+			}
+		}
+	}
+
+	for _, tag := range query.ExcludeTags {
+		tagLabel := []byte(strings.ToLower(string(tag)))
+		ids, err := b.tagsIndex.List(tx, tagLabel)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, k := range ids {
+			delete(jobSet, string(k))
+		}
+	}
+
+	var result []model.Job
+
+	bucket := tx.Bucket(BucketJobs)
+	for key := range jobSet {
+		var job model.Job
+		data := bucket.Get([]byte(key))
+		err := json.Unmarshal(data, &job)
+		if err != nil {
+			return nil, err
+		}
+		result = append(result, job)
+	}
+
+	listSorter := b.getListSorter(result, query)
+	sort.Slice(result, listSorter)
+
+	limit := query.Limit
+	if limit == 0 {
+		limit = len(result)
+	} else {
+		limit = math.Min(len(result), limit+query.Offset)
+	}
+
+	return result[query.Offset:limit], nil
+}
+
+func (b *BoltJobStore) getListSorter(jobs []model.Job, query jobstore.JobQuery) func(i, j int) bool {
+	return func(i, j int) bool {
+		switch query.SortBy {
+		case "id":
+			if query.SortReverse {
+				// what does it mean to sort by ID?
+				return jobs[i].Metadata.ID > jobs[j].Metadata.ID
+			} else {
+				return jobs[i].Metadata.ID < jobs[j].Metadata.ID
+			}
+		case "created_at":
+			if query.SortReverse {
+				return jobs[i].Metadata.CreatedAt.UTC().Unix() > jobs[j].Metadata.CreatedAt.UTC().Unix()
+			} else {
+				return jobs[i].Metadata.CreatedAt.UTC().Unix() < jobs[j].Metadata.CreatedAt.UTC().Unix()
+			}
+		default:
+			return false
+		}
+	}
+}
+
+// GetJobState returns the current job state for the provided job id
+func (b *BoltJobStore) GetJobState(ctx context.Context, jobID string) (model.JobState, error) {
+	var state model.JobState
+
+	err := b.database.View(func(tx *bolt.Tx) (err error) {
+		state, err = b.getJobState(tx, jobID)
+		return
+	})
+
+	return state, err
+}
+
+func (b *BoltJobStore) getJobState(tx *bolt.Tx, jobID string) (model.JobState, error) {
+	var state model.JobState
+
+	data := GetBucketData(tx, BucketPathState, []byte(jobID))
+	if data == nil {
+		return state, bacerrors.NewJobNotFound(jobID)
+	}
+
+	err := json.Unmarshal(data, &state)
+	return state, err
+}
+
+// GetInProgressJobs gets a list of the currently in-progress jobs
+func (b *BoltJobStore) GetInProgressJobs(ctx context.Context) ([]model.JobWithInfo, error) {
+	var infos []model.JobWithInfo
+	err := b.database.View(func(tx *bolt.Tx) (err error) {
+		infos, err = b.getInProgressJobs(tx)
+		return
+	})
+	return infos, err
+}
+
+func (b *BoltJobStore) getInProgressJobs(tx *bolt.Tx) ([]model.JobWithInfo, error) {
+	var infos []model.JobWithInfo
+	var keys [][]byte
+
+	keys, err := b.inProgressIndex.List(tx, BucketJobsInProgress)
+	if err != nil {
+		return nil, err
+	}
+
+	bktJobs := tx.Bucket(BucketJobs)
+	bktState, err := NewBucketPath(BucketPathState).Get(tx, false)
+	if err != nil {
+		return nil, err
+	}
+
```

Comment:
sometimes you are instantiating bucket directly and other times you are using bucket path, which can get a little confusing. What is the reason for not instantiating the state bucket directly here? The bucket should already exist as it was created in the constructor

Reply:
Have changed to be a bit more consistent.